flyteorg/flytepropeller
# Allow checkpoint resume when recovering a workflow

# TL;DR
This changes propagates task level checkpoint info between failed and recovered node executions. Previously intra task checkpointing was only supported for task level retries within a single node execution.

Related PR: https://github.com/flyteorg/flyteadmin/pull/479

**NOTE: cannot merge this until https://github.com/flyteorg/flytepropeller/pull/467 is resolved**

## Type

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
This change
* Saves checkpoint path from successful or failed task nodes in `TaskNodeMetadata`
* Sends to flyteadmin as part of the `NodeExecutionEvent`, to be stored in the db
* When attempting to recover a node execution, read the checkpoint path from `NodeExeuction.Closure` and store in the `ExecutableNodeStatus` (persisted to the CRD) so that it's available for later phase processing
* Provide this previous checkpoint path to the task on attempt 0, else continue passing path from attempt N-1 in the current node execution

## Testing
Added various unit tests

Verified with a local version of Flyte, following setup steps here. 
1. Run a task indefinitely that checkpoints every 10s (see code below)
2. Use kubectl to kill the pod (simulate infra failure)
3. Recover the workflow
4. On recovery, the task will find a checkpoint and exit successfully

```py3
import logging
import time

from flytekit import current_context, task, workflow

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__file__)


@task
def t():
    cp = current_context().checkpoint
    prev = cp.read()
    if prev:
        logger.info(f"Recovering from iteration {prev.decode()}")
        return
    iterations = 0
    while True:
        time.sleep(10)
        iterations += 1
        logger.info(iterations)
        cp.write(f"{iterations}".encode())


@workflow
def wf():
    t()
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2254

## Follow-up issue
Uncovered a few issues along the way
https://github.com/flyteorg/flyte/issues/2894
https://github.com/flyteorg/flytesnacks/pull/894 (PR)
https://github.com/flyteorg/flytekit/pull/1189 (PR)
